File: 17.letter-combinations-of-a-phone-number.py
# @leet imports start
from string import *
from re import *
from datetime import *
from collections import *
from heapq import *
from bisect import *
from copy import *
from math import *
from random import *
from statistics import *
from itertools import *
from functools import *
from operator import *
from io import *
from sys import *
from json import *
from builtins import *
import string
import re
import datetime
import collections
import heapq
import bisect
import copy
import math
import random
import statistics
import itertools
import functools
import operator
import io
import sys
import json
from typing import *
# @leet imports end

from typing import List

# @leet start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        strs = []
        m = {'2' : 'abc', '3' : 'def', '4': 'ghi', '5' : 'jkl', '6' : 'mno', '7' : 'pqrs', '8' : 'tuv', '9' : 'wxyz'}

        def letter(i, pref):
            print(f'{' ' * i}i={i} pref={pref}')
            if i == len(digits):
                print(f"out -> {pref}")
                strs.append(pref)
            else:
                for l in m[digits[i]]:
                    letter(i + 1, pref + l)
        letter(0, '')
        return strs

# @leet end

Solution().letterCombinations('23')

File: 17.letter-combinations-of-a-phone-number.cpp
// @leet imports start
#include <bits/stdc++.h>
#include <string>

using namespace std;
// @leet imports end


// @leet start
class Solution
{
public:
    vector<string> letterCombinations(const string &digits)
    {
        vector<string> strs = {};
        Comb(0, "", digits, strs);
        return strs;
        //vector<string> strs = {""};
        //for (char d : digits) {
        //    vector<string> strs2;
        //    for (const auto &s : strs) {
        //        for (char ch : M[d - '0']) {
        //            strs2.push_back(s + ch);
        //        }
        //    }
        //    strs.swap(strs2);
        //}
        //return strs;
    }

private:
    static constexpr const string_view M[] = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };

    void Comb(int i, std::string pref, const std::string &digits, vector<string> &out)
    {
        if (i == digits.size()) {
            out.push_back(std::move(pref));
            return;
        }

        for (char ch : M[digits[i] - '0']) {
            Comb(i + 1, pref + ch, digits, out);
        }
    }
};
// @leet end

int main()
{
    for (const auto &s : Solution{}.letterCombinations("23")) {
        cout << ' ' << s << endl;
    }
    return 0;
}


